security/pkg/pki/ca/ca.go

```diff
@@ -317,9 +329,244 @@ func (ca *IstioCA) SignWithCertChain(csrPEM []byte, subjectIDs []string, ttl tim
 
 // GetCAKeyCertBundle returns the KeyCertBundle for the CA.
 func (ca *IstioCA) GetCAKeyCertBundle() util.KeyCertBundle {
+	ca.mutex.RLock()
+	defer ca.mutex.RUnlock()
 	return ca.keyCertBundle
 }
 
+// setCAKeyCertBundle sets KeyCertBundle to the CA.
+func (ca *IstioCA) setCAKeyCertBundle(newBundle util.KeyCertBundle) {
+	ca.mutex.Lock()
+	defer ca.mutex.Unlock()
+	ca.keyCertBundle = newBundle
+}
+
+// SelfSignedCARootCertRotationConfig configs the automatic root certificate
+// rotation for self-signed CA instance.
+type SelfSignedCARootCertRotationConfig struct {
+	CheckInterval       time.Duration
+	CaCertTTL           time.Duration
+	RetryInterval       time.Duration
+	GracePeriodRatio    float32
+	Client              corev1.CoreV1Interface
+	CaStorageNamespace  string
+	DualUse             bool
+	ReadSigningCertOnly bool
+	Org                 string
+	RootCertFile        string
+	Metrics             monitoring.MonitoringMetrics
+}
+
+// RotateRootCert refreshes root certs and updates config map accordingly.
+func (ca *IstioCA) RotateRootCert(config *SelfSignedCARootCertRotationConfig) {
+	ticker := time.NewTicker(config.CheckInterval)
+	for {
+		select {
+		case <-ticker.C:
+			ca.checkAndRotateRootCert(config)
+		case _, ok := <-ca.StopRotateJob:
+			if !ok {
+				log.Info("Self-signed CA is shutting down, stop root cert rotation job")
```

Comment:
I just keep this consistent as other controllers, the controllers create goroutines and use a channel to shutdown those goroutines. When we add unit tests, we also need to shutdown goroutine when one test is finished, so that the next test is not affected by leftover goroutine.